### PR TITLE
Integrate LM Eval Harness CLI

### PR DESCRIPTION
**Pull request by eric-czech**
refs: https://github.com/kuleshov-group/dllm-dev/issues/5

This adds a CLI wrapper over https://github.com/EleutherAI/lm-evaluation-harness.  That wrapper is essentially a passthrough right now for a config like this: [configs/eval/lm_eval_harness/cli.yaml](https://github.com/kuleshov-group/dllm-dev/blob/11a1f9f1d5559c8da73e3abcc2d9bc91aa8e211c/configs/eval/lm_eval_harness/cli.yaml).

This makes it possible to run a simple, test eval this way (on CPU):

```bash
python -m scripts.eval.lm_eval_harness \
  args.model_args="pretrained\=distilgpt2" \
  args.device=cpu
```

Or a more realistic eval (i.e. `LLaDA 8B`) with a GPU this way:

```bash
python -m scripts.eval.lm_eval_harness \
  args.tasks="arc_easy,hellaswag,winogrande" \
  args.model_args="pretrained\=GSAI-ML/LLaDA-8B-Base" \
  args.limit=250 \
  args.device=cuda
# hf (pretrained=GSAI-ML/LLaDA-8B-Base,trust_remote_code=True), gen_kwargs: (None), limit: 250.0, num_fewshot: 0, batch_size: 1
# |  Tasks   |Version|Filter|n-shot| Metric |   |Value|   |Stderr|
# |----------|------:|------|-----:|--------|---|----:|---|-----:|
# |arc_easy  |      1|none  |     0|acc     |↑  |0.316|±  |0.0295|
# |          |       |none  |     0|acc_norm|↑  |0.276|±  |0.0283|
# |hellaswag |      1|none  |     0|acc     |↑  |0.256|±  |0.0277|
# |          |       |none  |     0|acc_norm|↑  |0.232|±  |0.0268|
# |winogrande|      1|none  |     0|acc     |↑  |0.480|±  |0.0317|
```

The CLI itself is documented here: [EleutherAI/lm-evaluation-harness/docs/interface.md](https://github.com/EleutherAI/lm-evaluation-harness/blob/main/docs/interface.md).

I tried to run this on a E2D2 checkpoint like [kuleshov-group/e2d2-gsm8k-finetune-Qwen3-2B](https://huggingface.co/kuleshov-group/e2d2-gsm8k-finetune-Qwen3-2B), but it needs to implement `AutoModelForCausalLM` rather than `AutoModelForMaskedLM` as LLaDA does (e.g. [here](https://huggingface.co/GSAI-ML/LLaDA-8B-Base/blob/main/config.json)).  Masked models are not apparently supported.

---

<details><summary>Commands</summary>

### Local 

Local testing on CPU:

```bash
# Run the CLI directly with a small model
python -m lm_eval --model hf --model_args pretrained=distilgpt2 --tasks arc_easy \
--batch_size 1 --num_fewshot 0 --device cpu \
--output_path ./test_output --log_samples --seed 42 --limit 3

# Run the config + CLI wrapper
source .venv/bin/activate && python -m scripts.eval.lm_eval_harness
```

### Remote

Remote execution w/ GPU:

```bash
# Create cluster
sky launch --infra lambda --cluster dllm --gpus A100 --disk-size 100 --env HUGGING_FACE_HUB_TOKEN --env WANDB_API_KEY skypilot/cluster.sky.yaml

# Sync code
rsync -Pzr . ubuntu@dllm:/home/ubuntu/sky_workdir --exclude '.venv' --exclude '.git' --exclude '__pycache__' --exclude 'outputs/'

# Connect and run container
ssh dllm; cd sky_workdir
sudo docker run --rm -it \
--env-file ~/.env \
-e RUN_DIR=/workspace/outputs \
-e DATA_DIR=/workspace/data \
-e NUM_NODES=1 \
-e WORLD_SIZE=1 \
-e NODE_RANK=0 \
-e HF_HOME=/workspace/.hf_cache \
-e TORCHINDUCTOR_CACHE_DIR=/workspace/.ti_cache \
-e TORCHINDUCTOR_AUTOGRAD_CACHE=1 \
-e TORCHINDUCTOR_FX_GRAPH_CACHE=1 \
-e PYTHONPATH=/workspace:/workspace/.hf_cache/modules \
--network host \
--shm-size=16g \
--gpus all --rm -v $(pwd):/workspace \
eczech/dllm-dev:py311-cuda128-torch270
```

```
# Run a simple test
python -m scripts.eval.lm_eval_harness \
  args.model_args="pretrained\=distilgpt2" \
  args.limit=10 args.device=cuda
hf (pretrained=distilgpt2), gen_kwargs: (None), limit: 10.0, num_fewshot: 0, batch_size: 1
|  Tasks   |Version|Filter|n-shot| Metric |   |Value|   |Stderr|
|----------|------:|------|-----:|--------|---|----:|---|-----:|
|arc_easy  |      1|none  |     0|acc     |↑  |  0.4|±  |0.1633|
|          |       |none  |     0|acc_norm|↑  |  0.2|±  |0.1333|
|hellaswag |      1|none  |     0|acc     |↑  |  0.3|±  |0.1528|
|          |       |none  |     0|acc_norm|↑  |  0.4|±  |0.1633|
|winogrande|      1|none  |     0|acc     |↑  |  0.6|±  |0.1633|

# Run a LLaDA eval
python -m scripts.eval.lm_eval_harness \
  args.model_args="pretrained\=GSAI-ML/LLaDA-8B-Base" \
  args.limit=250 \
  args.device=cuda
hf (pretrained=GSAI-ML/LLaDA-8B-Base,trust_remote_code=True), gen_kwargs: (None), limit: 250.0, num_fewshot: 0, batch_size: 1
|  Tasks   |Version|Filter|n-shot| Metric |   |Value|   |Stderr|
|----------|------:|------|-----:|--------|---|----:|---|-----:|
|arc_easy  |      1|none  |     0|acc     |↑  |0.316|±  |0.0295|
|          |       |none  |     0|acc_norm|↑  |0.276|±  |0.0283|
|hellaswag |      1|none  |     0|acc     |↑  |0.256|±  |0.0277|
|          |       |none  |     0|acc_norm|↑  |0.232|±  |0.0268|
|winogrande|      1|none  |     0|acc     |↑  |0.480|±  |0.0317|
```

To hack on the CLI itself:

```
 vi /opt/conda/lib/python3.11/site-packages/lm_eval/__main__.py
```

</details>

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -11,6 +11,10 @@ mosaicml-streaming==0.13.0
 pre-commit
 ruff
 sentencepiece==0.2.1
-torch==2.5.1
+# torch==2.5.1
+# Upgrade torch due to:
+#   ╰─▶ Because mosaicml==0.32.1 depends on torch>=2.6.0,<2.7.1 and you require mosaicml==0.32.1, we can conclude that you require torch>=2.6.0,<2.7.1.
+#       And because you require torch==2.5.1, we can conclude that your requirements are unsatisfiable.
+torch==2.7.0
```
Comment (eric-czech) [inline requirements.txt]: I think I saw this inconsistency before in the environment and forgot to fix it.  @yair-schiff / @mariannearr do you say any reason not to use torch 2.7.0 from now on?  Torch 2.5.1 is too old for composer 0.32.1.

Comment (yair-schiff) [reply requirements.txt]: Let's upgrade

**Comment by yair-schiff** [comment]
@eric-czech in the previous setup, I was using this wrapper class from `lm_eval_harness`:
```
from lm_eval.api.model import LM
```
and the idea was to implement the 4 required methods in the wrapper, utilizing the underlying model's apis to get the necessary values. The only assumption i was making on the pretrained models then is that they defined a `.generate` method.

In this new setup, you are assuming that the underlying pretrained model implements all the required methods by lm-eval harness, e.g. get likelihood, get rolling likelihood, etc. 

Is this a correct assessment of this PR?
